src/wallet/wallet2.cpp

```diff
@@ -1651,17 +1654,187 @@ void wallet2::scan_tx(const std::vector<crypto::hash> &txids)
+{
+  LOG_PRINT_L0("Processing " << sorted_txs_to_scan.tx_entries.size() << " txs, re-processing "
+      << sorted_txs_to_reprocess.tx_entries.size() << " txs");
+
   // Process the transactions in chronologically ascending order
-  while(!txq.empty()) {
-    auto& tx_info = txq.top();
+  std::vector<cryptonote::COMMAND_RPC_GET_TRANSACTIONS::entry> process_txs = merge_sorted_tx_entries(sorted_txs_to_scan.tx_entries, sorted_txs_to_reprocess.tx_entries);
+  for (const auto &tx_info : process_txs)
+  {
     cryptonote::transaction tx;
     crypto::hash tx_hash;
     THROW_WALLET_EXCEPTION_IF(!get_pruned_tx(tx_info, tx, tx_hash), error::wallet_internal_error, "Failed to get transaction from daemon (2)");
-    process_new_transaction(tx_hash, tx, tx_info.output_indices, tx_info.block_height, 0, tx_info.block_timestamp, false, tx_info.in_pool, tx_info.double_spend_seen, {}, {});
-    txq.pop();
+
+    // Ignore callbacks when re-processing a tx to avoid confusing feedback to user
+    bool ignore_callbacks = tx_hashes_to_reprocess.find(tx_hash) != tx_hashes_to_reprocess.end();
+    process_new_transaction(tx_hash, tx, tx_info.output_indices, tx_info.block_height, 0, tx_info.block_timestamp, cryptonote::is_coinbase(tx), tx_info.in_pool, tx_info.double_spend_seen, {}, {}, ignore_callbacks);
+
+    // Re-set destination addresses if they were previously set
+    if (m_confirmed_txs.find(tx_hash) != m_confirmed_txs.end() &&
+        dbd.detached_confirmed_txs_dests.find(tx_hash) != dbd.detached_confirmed_txs_dests.end())
+    {
+      m_confirmed_txs[tx_hash].m_dests = std::move(dbd.detached_confirmed_txs_dests[tx_hash]);
+    }
+  }
+
+  LOG_PRINT_L0("Done processing " << sorted_txs_to_scan.tx_entries.size() << " txs and re-processing "
+      << sorted_txs_to_reprocess.tx_entries.size() << " txs");
+}
+//----------------------------------------------------------------------------------------------------
+void reattach_blockchain(hashchain &blockchain, wallet2::detached_blockchain_data &dbd)
+{
+  if (!dbd.detached_blockchain.empty())
+  {
+    LOG_PRINT_L0("Re-attaching " << dbd.detached_blockchain.size() << " blocks");
+    for (size_t i = 0; i < dbd.detached_blockchain.size(); ++i)
+      blockchain.push_back(dbd.detached_blockchain[i]);
+  }
+
+  THROW_WALLET_EXCEPTION_IF(blockchain.size() != dbd.original_chain_size,
+    error::wallet_internal_error, "Unexpected blockchain size after re-attaching");
+}
+//----------------------------------------------------------------------------------------------------
+bool has_unexpected_tx_above_height(uint64_t height, const std::unordered_set<crypto::hash> &expected_txids, const wallet2::transfer_container &transfers,
+    const wallet2::payment_container &payments, const serializable_unordered_map<crypto::hash, wallet2::confirmed_transfer_details> &confirmed_txs)
+{
+  for (const auto &td : transfers)
+    if (td.m_block_height >= height && expected_txids.find(td.m_txid) == expected_txids.end())
```

Comment:
This has >= for a function named "above", intended ?

Reply:
Renamed to `has_unexpected_tx_at_height_or_above`